plaidml/bridge/openvino/ops/detection_output.cpp

```diff
@@ -0,0 +1,223 @@
+    }
+
+    variances = op::squeeze(variances, {1});
+    priors = op::squeeze(priors, {1});
+
+    edsl::Tensor decoded_bboxes;
+    if (with_add_pred) {
+      // Update confidence if there are 5 inputs.
+      Tensor IX = edsl::index({edsl::TensorDim(num_priors)}, 0);
+      IX = IX * 2 + 1;
+      edsl::Tensor arm_conf = edsl::gather(ArmConfidence, IX).axis(1);
+      arm_conf = op::repeat(op::unsqueeze(arm_conf, {-1})).count(num_classes).axis(2);
+      confidence = edsl::select(arm_conf < objectness_score, cast(edsl::Tensor{0.0f}, DType::FLOAT32), confidence);
+
+      // Decode bouding boxes
+      edsl::Tensor arm_loc = edsl::reshape(ArmLocation, location_shape);
+      auto decoded_priors =
+          DecodeBoxes(priors, variances, arm_loc, code_type, i_h, i_w, batch, num_priors, clip_before_nms);
+      decoded_bboxes =
+          DecodeBoxes(decoded_priors, variances, location, code_type, i_h, i_w, batch, num_priors, clip_before_nms);
+    } else {
+      decoded_bboxes =
+          DecodeBoxes(priors, variances, location, code_type, i_h, i_w, batch, num_priors, clip_before_nms);
+    }
+
+    // Transpose the confidence to match the input shape of `scores` in NMS.
+    // confidence -> {batch, num_boxes, num_classes}
+    // scores -> {batch, num_classes, num_boxes}
+    edsl::Tensor nms_conf = op::transpose(confidence, edsl::make_tuple<int64_t>({0, 2, 1}));
+    // Remove the background class.
+    if (background_label_id > -1) {
+      std::vector<edsl::Tensor> idx_vec;
+      edsl::Tensor idx_first_half = edsl::index({edsl::TensorDim(background_label_id)}, 0);
+      edsl::Tensor idx_second_half =
+          op::slice(edsl::index({edsl::TensorDim(num_classes)}, 0)).add_dim(background_label_id + 1, num_classes);
+      idx_vec.push_back(idx_first_half);
+      idx_vec.push_back(idx_second_half);
+      edsl::Tensor gather_idxs = background_label_id == 0 ? idx_second_half : op::concatenate(idx_vec, 0);
+      nms_conf = edsl::gather(nms_conf, gather_idxs).axis(1);
+    }
+
+    edsl::Tensor iou_threshold = cast(edsl::Tensor{nms_threshold}, DType::FLOAT32);
+    edsl::Tensor score_threshold = cast(edsl::Tensor{confidence_threshold}, DType::FLOAT32);
+    std::vector<edsl::Tensor> result = op::nms(decoded_bboxes, nms_conf, iou_threshold, score_threshold, top_k)
+                                           .soft_nms_sigma(0.0f)
+                                           .center_point_box(false)
+                                           .sort_result_descending(false)
+                                           .box_output_type(DType::INT32)
+                                           .build();
+    edsl::Tensor selected_indices = result[0];
+    auto selected_indices_shape = selected_indices.compute_shape().sizes();
+    edsl::Tensor selected_scores = result[1];
+    edsl::Tensor valid_outputs = result[2];
+
+    edsl::Tensor batch_slice = op::slice(selected_indices).add_dim(0, selected_indices_shape[0]).add_dim(0, 1);
```

Comment:
Can we use gather directly for those ***_slice and get gather_idxs directlly?

Reply:
Do you suggest using one `gather` instead of two `op::slice` here?

Reply:
Done.